# <span style="color:rgb(160,0,86)">Datenbeschreibung mit Python</span>

***

## <span style="color:rgb(160,0,86)">Lernziele</span>

- Sie können absolute und relative Häufigkeiten und Summenhäufigkeiten berechnen und verstehen ihre Bedeutung.
- Sie kennen Lagemasse und Streumasse für nominale, ordinale und metrische Merkmale und können sie berechnen.  

***

### <span style="color:rgb(160,0,86)">Was ist eine Häufigkeitsanalyse?</span>

Die grundlegendste Aufbereitung statistischer Daten ist die Darstellung, wie häufig die Werte $w$ eines Merkmals vorkommen. 

- Die **absolute Häufigkeit** $\;\pmb{n(w)}\;$ ist die Anzahl, wie oft ein Wert $w$ vorkommt.
- Die **relative Häufigkeit** $\;\pmb{h(w)=\displaystyle\frac{n(w)}{n}}\;$ ist der Anteil, mit dem ein Wert $w$ vorkommt. 

Weil *oridinale* und *metrische* Merkmalen eine Reihenfolge haben, können wir Häufigkeiten auch der Reihe nach kumulieren:

- Die **absolute Summenhäufigkeit** $\;\pmb{N(w)}\;$ ist die Anzahl, wie oft Werte kleiner oder gleich $w$ vorkommen.
- Die **relative Summenhäufigkeit** $\;\pmb{H(w)=\displaystyle\frac{N(w)}{n}}\;$ ist der Anteil, mit dem Werte kleiner oder gleich $w$ vorkommen.

Die Auflistung aller Werte $w_1, w_2, \ldots, w_m$ eines Merkmals zusammen mit ihren Häufigkeiten nennen wir eine **empirische Häufigkeitsverteilung**.

***Merke:*** Aus der empirischen Häufigkeitsverteilung können wir nicht mehr herauslesen, was für einen Wert eine statistische Einheit bei einem Merkmal hat. Wir sehen nur noch, wie oft oder wie häufig die Werte eines Merkmals vorkommen. **Es geht Information verloren!** 

***

#### <span style="color:rgb(160,0,86)">Beispiel:</span>

Betrachten wir 10 ehemalige Studierende mit den Merkmalen **Geschlecht**, **Gesamtprädikat** und **Alter**:

$$\begin{bmatrix}
\text{Anna}&w&\text{ausgezeichnet}&27\\
\text{Beat}&m&\text{gut}&34\\
\text{Cary}&m&\text{sehr gut}&29\\
\text{Dana}&w&\text{sehr gut}&24\\
\text{Elif}&w&\text{gut}&25\\
\text{Faro}&m&\text{ausgezeichnet}&27\\
\text{Gabi}&w&\text{sehr gut}&27\\
\text{Hans}&m&\text{genügend}&69\\
\text{Ivea}&w&\text{sehr gut}&26\\
\text{Jose}&w&\text{gut}&31
\end{bmatrix}$$ 

Diese Daten speichern wir in ein **DataFrame mit Namen df**:

In [1]:
namen = ["Anna","Beat","Cary","Dana","Elif","Faro","Gabi","Hans","Ivea","Jose"]
geschlecht = ["w","m","m","w","w","m","w","m","w","w"]
prädikat = ["ausgezeichnet","gut","sehr gut","sehr gut","gut","ausgezeichnet","sehr gut","genügend","sehr gut","gut"]
alter = [27,34,29,24,25,27,27,69,26,31]

import pandas as pd
df = pd.DataFrame({"Name":namen,
                   "Geschlecht":geschlecht,
                   "Prädikat":prädikat,
                   "Alter":alter})
df

,Name,Geschlecht,Prädikat,Alter
0,Anna,w,ausgezeichnet,27
1,Beat,m,gut,34
2,Cary,m,sehr gut,29
3,Dana,w,sehr gut,24
4,Elif,w,gut,25
5,Faro,m,ausgezeichnet,27
6,Gabi,w,sehr gut,27
7,Hans,m,genügend,69
8,Ivea,w,sehr gut,26
9,Jose,w,gut,31


Das **Geschlecht** ist ein **nominales Merkmal**. 

Die absolute empirische Verteilung finden wir mit der Funktion **np.unique**, wenn wir **return_counts=True** setzen. 

Die Funktion gibt **zwei Listen** zurück, die wir direkt in den zwei Variablen **werte**, **absH** speichern können:   

In [2]:
import numpy as np

werte, absH = np.unique(df["Geschlecht"],return_counts=True)

Nun können wir die **Werte und ihre Häufigkeiten** ausgeben:

In [3]:
print("Verteilung der absoluten Häufigkeiten:")
for i in range(len(werte)):
    print("  n(" + werte[i] + ") =",absH[i])
print("Verteilung der relativen Häufigkeiten:")
for i in range(len(werte)):
    print("  h(" + werte[i] + ") =",absH[i]/df.shape[0])

Verteilung der absoluten Häufigkeiten:
  n(m) = 4
  n(w) = 6
Verteilung der relativen Häufigkeiten:
  h(m) = 0.4
  h(w) = 0.6


Das **Prädikat** ist ein **ordinales Merkmal**. Seine empirische Verteilung finden wir analog:

In [4]:
werte, absH = np.unique(df["Prädikat"],return_counts=True)
werte

array(['ausgezeichnet', 'genügend', 'gut', 'sehr gut'], dtype=object)

In der Ausgabe der Werte stimmt die **natürliche Reihenfolge** nicht - die Werte werden alphabetisch gelistet! 

Wenn wir aber einen **for-loop über eine Liste** mit Indizes in richtiger Reihenfolge machen, erhalten wir die natürliche Anordnung: 

In [21]:
print("Verteilung der absoluten Häufigkeiten:")
for i in [1,2,3,0]:
    print("  n(" + werte[i] + ") =",absH[i])
print("Verteilung der relativen Häufigkeiten:")
for i in [1,2,3,0]:
    print("  h(" + werte[i] + ") =",absH[i]/df.shape[0])

Verteilung der absoluten Häufigkeiten:
  n(gut) = 1
  n(sehr gut) = 3
  n(ausgezeichnet) = 4
  n(genügend) = 2
Verteilung der relativen Häufigkeiten:
  h(gut) = 0.1
  h(sehr gut) = 0.3
  h(ausgezeichnet) = 0.4
  h(genügend) = 0.2


Für **ordinale Merkmale** können wir auch die **Summenhäufigkeiten** berechnen:

In [22]:
print("Verteilung der absoluten Summenhäufigkeiten:")
absSH = 0
for i in [1,2,3,0]:
    absSH += absH[i]
    print("  N(" + werte[i] + ") =",absSH)
print("Verteilung der relativen Summenhäufigkeiten:")
absSH = 0
for i in [1,2,3,0]:
    absSH += absH[i]
    print("  H(" + werte[i] + ") =",absSH/df.shape[0])

Verteilung der absoluten Summenhäufigkeiten:
  N(gut) = 1
  N(sehr gut) = 4
  N(ausgezeichnet) = 8
  N(genügend) = 10
Verteilung der relativen Summenhäufigkeiten:
  H(gut) = 0.1
  H(sehr gut) = 0.4
  H(ausgezeichnet) = 0.8
  H(genügend) = 1.0


Das **Alter** ist ein **metrisches Merkmal**. Seine empirische Verteilung finden wir analog:

In [7]:
werte, absH = np.unique(df["Alter"],return_counts=True)

print("Verteilung der absoluten Häufigkeiten:")
for i in range(len(werte)):
    print("  n(" + str(werte[i]) + ") =",absH[i])
print("Verteilung der relativen Häufigkeiten:")
for i in range(len(werte)):
    print("  h(" + str(werte[i]) + ") =",absH[i]/df.shape[0])

Verteilung der absoluten Häufigkeiten:
  n(24) = 1
  n(25) = 1
  n(26) = 1
  n(27) = 3
  n(29) = 1
  n(31) = 1
  n(34) = 1
  n(69) = 1
Verteilung der relativen Häufigkeiten:
  h(24) = 0.1
  h(25) = 0.1
  h(26) = 0.1
  h(27) = 0.3
  h(29) = 0.1
  h(31) = 0.1
  h(34) = 0.1
  h(69) = 0.1


**Metrische Merkmale** nehmen oft *viele verschiedene Zahlenwerte* an. Daher ist es praktikabler, wenn wir eine Funktion definieren, um die **Summenhäufigkeiten**  $\;\pmb{N(w)}\;$ und $\;\pmb{H(w)}\;$ abzufragen:

In [8]:
def Summenhäufigkeit(w,n,x):
    ''' w: Auflistung der eindeutigen Werte als numpy array
        n: Auflistung der entsprechenden abs. Häufigkeiten als numpy array
        x: Ein Alter als integer '''
    N = 0
    i = 0
    while i < len(w) and w[i] <= x:
        N += n[i]
        i += 1
    ''' return: (Anzahl Jahre <= x, Anteil Jahre <= x) '''
    return (N,N/sum(n))
    
Summenhäufigkeit(werte,absH,25)

(np.int64(2), np.float64(0.2))

### <span style="color:rgb(160,0,86)">Was sind Lagemasse und Streumasse?</span>

Die empirische Häufigkeitsverteilung beantwortet die Frage: Wie oft kommen die Werte eines Merkmals vor? Mit **statistischen Masszahlen** sollen nun spezifische Informationen aus den Daten herausgefiltert werden.      

- Ein **Lagemass** beschreibt mit einem typischen Wert, ***wo*** die Ausprägungen eines Merkmals liegen.
- Ein **Streumass** beschreibt mit einem typischen Wert, ***wie unterschiedlich***  die Ausprägungen eines Merkmals sind.

#### <span style="color:rgb(160,0,86)">Für nominale Merkmale:</span>
- ***Lagemass:*** Weil die Werte keine natürliche Reihenfolge haben, können wir keine *Mitte* festlegen. Als repräsentativer Wert für die Lage der Werte wird oft ***der häufigste Wert***, der sogenannte **Modus** $\;\bar{x}_{D}\;$ angegeben.
- ***Streumass:*** Wir wollen angeben, wie gleichmässig sich die ungeordneten Werte verteilen. Wenn bei einem Merkmal nur ein Wert vorkommt, gibt es keine Streuung. Wenn hingegen alle Werte eines Merkmals gleich häufig auftreten, gibt es maximale Unterschiedlichkeit. Als Mass für diese Charakteristik wird häufig der **Dispersionsindex** $$P = \displaystyle\frac{m}{m-1}\big(h(w_1)\cdot(1-h(w_1))+h(w_2)\cdot(1-h(w_2))+\ldots+h(w_m)\cdot(1-h(w_m)\big)$$ gebraucht. Wenn $\,P>0.9\,$ ist, haben wir starke Dispersion und wenn $\,P<0.8\,$ ist, haben wir geringe Dispersion. <br>
Wenn im Extremfall **ein Wert eine relative Häufigkeit von 1.0** hat, dann ist $\;\pmb{P=0}\,$:
$$P = \displaystyle\frac{m}{m-1}\big(\,\underbrace{\textcolor{red}{1}\cdot(1-\textcolor{red}{1})}_{=\,0}+\underbrace{\textcolor{red}{0}\cdot(1-\textcolor{red}{0})}_{=\,0}\;+\;\ldots\;+\;\underbrace{\textcolor{red}{0}\cdot(1-\textcolor{red}{0})}_{=\,0}\,\big)$$  Wenn **alle Werte die gleiche relative Häufigkeit** haben, dann ist $\;\pmb{P=1}\,$:
$$P = \displaystyle\frac{m}{m-1}\big(\,\underbrace{\textcolor{red}{\textstyle\frac{1}{m}}\cdot(1-\textcolor{red}{\textstyle\frac{1}{m}})}_{=\,\frac{m-1}{m^2}}+\underbrace{\textcolor{red}{\textstyle\frac{1}{m}}\cdot(1-\textcolor{red}{\textstyle\frac{1}{m}})}_{=\,\frac{m-1}{m^2}}\;+\;\ldots\;+\;\underbrace{\textcolor{red}{\textstyle\frac{1}{m}}\cdot(1-\textcolor{red}{\textstyle\frac{1}{m}})}_{=\,\frac{m-1}{m^2}}\,\big)$$ 

Die Methode **mode()** liefert den **Modus** $\;\bar{x}_{D}\;$ direkt:
- Gibt es **genau einen** Wert mit der höchsten Häufigkeit, wird dieser Wert zurückgegeben.
- Gibt es **mehrere** Werte mit gleicher Häufigkeit, werden alle diese Werte zurückgegeben.

Für das Merkmal **Geschlecht** gilt:

In [9]:
df["Geschlecht"].mode() # ein Series Objekt mit allen Werten

0    w
Name: Geschlecht, dtype: object

In [ ]:
modus = df["Geschlecht"].mode()[0] # nur ein Wert ist am häufigsten
print("x_D =",modus)

x_D = w


Den **Disperesionsindex** $\;\pmb{P}\;$ können wir wieder mit der Funktion **np.unique()** berechnen. 

Für das Merkmal **Geschlecht** gilt: 

In [11]:
werte, absH = np.unique(df["Geschlecht"],return_counts=True)

m = len(werte)
relH = absH/df.shape[0]

# Dispersionsindex berechnen:
P = m/(m-1) * sum(relH * (1 - relH)) # Broadcasting!

print("P =",P)

P = 0.96


***
Das selbe **von Hand**, ohne Python:

Aus den relativen Häufigkeiten $\;\texttt{[}
\overset{\large 0.4}{\texttt{"m"}},
\overset{\large 0.6}{\texttt{"w"}}\texttt{]}\;$ erkennen wir den **Modus** $\;\bar{x}_{D}=\texttt{"w"}\;$ und finden den **Dispersionsindex** $$P = \displaystyle\frac{2}{2-1}\big(0.4\cdot(1-0.4)+0.6\cdot(1-0.6)\big)=0.96\,.$$
***

#### <span style="color:rgb(160,0,86)">Für ordinale Merkmale:</span>
- ***Lagemass:*** Die Werte haben eine natürliche Reihenfolge, wir können aber mit diesen Werten im allgemeinen nicht rechnen. Wenn wir die Daten aufsteigend sortieren, können wir das sogenannte $p$**-Quantil** bestimmen. Das ist ein Wert $\,\pmb{x_p}\,$ des Merkmals, der die Daten in einen **unteren** und einen **oberen Teil** zerlegt:

  Im ***unteren Teil*** sind $\,p\%\,$ der Daten kleiner oder gleich als der Wert $\,\pmb{x_p}\,$


  Im ***oberen Teil*** sind $\,(1-p)\%\,$ der Daten grösser oder gleich als der Wert $\,\pmb{x_p}\,$.
  
  Wir wählen den **Werte** $\,\pmb{w}\,$ für das $p$-**Quantil** $\,\pmb{x_p}\,$ so, dass zum ersten Mal die relative Summenhäufigkeit $H(w)\geqslant p$ ist. Für alle Werte $w$ vor $x_p$ muss daher $H(w)<p$ gelten.

  Der Wert $\pmb{x_{50}}$ heisst **Median**, weil in beiden Teilen ungefähr gleich viele Daten liegen. <br>
  Der Wert $\pmb{x_{25}}$ heisst **unteres Quartil**, weil ungefähr ein viertel der Daten kleiner als $\pmb{x_{25}}$ sind. <br>
  Der Wert $\pmb{x_{75}}$ heisst **oberes Quartil**, weil ungefähr drei viertel der Daten kleiner als $\pmb{x_{75}}$ sind.
       
- ***Streumass:*** Wegen der natürlichen Reihenfolge können wir die Streuung differnzierter beschreiben. **Streuung ist gross**, wenn die Werte ***gleichmässig beim kleinsten*** und ***beim grössten*** Wert des Merkmals liegen. Als Mass für diese Charakteristik wird häufig die **Diversität** $$D = \displaystyle\frac{4}{m-1}\big(H(w_1)\cdot(1-H(w_1))+H(w_2)\cdot(1-H(w_2))+\ldots+H(w_m)\cdot(1-H(w_m)\big)$$ gebraucht. Wenn $D>0.8$ ist, haben wir starke Diversität und wenn $D<0.6$ ist, haben wir geringe Diversität.<br>
Wenn im Extremfall **ein Wert eine relative Häufigkeit von 1.0** hat, dann ist $\;\pmb{D=0}\,$: $$D = \displaystyle\frac{4}{m-1}\big(\,\underbrace{\textcolor{red}{0}\cdot(1-\textcolor{red}{0})}_{=\,0}+\underbrace{\textcolor{red}{1}\cdot(1-\textcolor{red}{1})}_{=\,0}\;+\;\ldots\;+\;\underbrace{\textcolor{red}{1}\cdot(1-\textcolor{red}{1})}_{=\,0}\,\big)$$ Wenn **50% der Daten beim kleinsten und 50% beim grössten** Wert liegen, dann ist $\;\pmb{D=1}\,$: $$D = \displaystyle\frac{4}{m-1}\big(\,\underbrace{\textcolor{red}{\textstyle\frac{1}{2}}\cdot(1-\textcolor{red}{\textstyle\frac{1}{2}})}_{=\,\frac{1}{4}}\;+\;\ldots\;+\;\underbrace{\textcolor{red}{\textstyle\frac{1}{2}}\cdot(1-\textcolor{red}{\textstyle\frac{1}{2}})}_{=\,\frac{1}{4}}+\underbrace{\textcolor{red}{1}\cdot(1-\textcolor{red}{1})}_{=\,0}\,\big)$$  

Für das ordinale Merkmal **Prädikat** können auch den **Modus** mit der Methode **mode()** bestimmen: 

In [12]:
modus = df["Prädikat"].mode()[0]
print("x_D =",modus)

x_D = sehr gut


Wegen der natürlichen Ordnung ist der **Median** $\,\pmb{x_{50}}\,$ zur Beschreibung der mittleren Lage besser geeignet. Wieder mit der Methode **unique()** finden wir:  

In [ ]:
werte, absH = np.unique(df["Prädikat"],return_counts=True)
m = len(werte) # Anzahl verschiedene Werte

# Werte in die natürliche Reihenfolge umspeichern:
werte = np.array([werte[1],werte[2],werte[3],werte[0]])
relH = np.array([absH[1],absH[2],absH[3],absH[0]])/df.shape[0]

# Relative Summenhäufigkeiten bestimmen (Anteil mmit dem Werte kleiner
# oder gleich vorkommen)
relSH = np.array([relH[0],sum(relH[:2]),sum(relH[:3]),sum(relH)])

# Median suchen:
i = 0
while i < m:
    if relSH[i] > 0.5: # Welcher Wert ist bei 50%
        break
    i += 1
median = werte[i]

print("Median =",median)

Median = sehr gut


Mit den relativen Summenhäufigkeiten finden wir auch sofort die **Diversität** $\,\pmb{D}\,$ für das Merkmal **Prädikat**:

In [14]:
D = 4/(m-1) * sum(relSH * (1-relSH)) # Broadcasting!
print("D =",D)

D = 0.6533333333333333


***
Das selbe **von Hand**, ohne Python:

Aus den relativen Summenhäufigkeiten $$\texttt{[}
\overset{\large 0.1}{\texttt{"genügend"}},
\overset{\large 0.4}{\texttt{"gut"}},
\overset{\large 0.8}{\texttt{"sehr gut"}}, 
\overset{\large 1.0}{\texttt{""ausgezeichnet"}}\texttt{]}$$ erkennen wir den **Median** $\,\pmb{x_{50} = \texttt{"sehr gut"}}\,$ und finden die **Diversität** $$D = \displaystyle\frac{4}{4-1}\big(0.1\cdot(1-0.1)+0.4\cdot(1-0.4)+0.8\cdot(1-0.8)+1.0\cdot(1-1.0)\big)=0.65\bar{3}\,.$$
***

#### <span style="color:rgb(160,0,86)">Für metrische Merkmale:</span>
- ***Lagemass:*** Da wir mit den Werten rechnen können, lässt sich die Summe aller Daten $x_1,x_2,\ldots,x_n$ **gleichmässig auf alle statistischen Einheiten verteilen**. Dieses **arithmetische Mittel** $\,\pmb{\bar{x}}\,$ beschreibt, wie gross die Werte im Mittel sind: $$\bar{x} = \frac{1}{n}(x_1+x_2+\ldots+x_n)=h(w_1)\cdot w_1+h(w_2)\cdot w_2 +\ldots + h(w_m)\cdot w_m$$ Das arithmetische Mittel wird aber von einzelnen extremen Werten stark beeinflusst. Der Median hingegen wird von extremen Werten weniger beeinflusst, weil sehr grosse oder sehr kleine Werte die Reihenfolge in der Mitte nicht verändern. Wenn ein Mass auf extreme Werte nicht oder nur moderat reagiert, heisst es **robust**. 
- ***Streumass:*** Bei metrischen Daten sind die Quartile auch Zahlen. Daher können wir den **Qartilsabstand** $\,\pmb{Q = x_{75}-x_{25}}\,$ vom unteren Quartil zum oberen Quartil berechnen. In diesem Bereich liegen ungefähr **50\% der mittleren Daten**. Wenn also $Q$ klein ist, dann liegen die Daten eng beieinander und wenn $Q$ gross ist, dann streuen die Daten stark. Der Quartilsabstand wird durch extreme Werte nicht stark beeinflusst, er ist also auch **robust**.<br>
Eine anderes Mass für die Streuung ist die **mittlere Abweichung** $\,\pmb{MD}\,$ der Daten vom arithmetischen Mittel $$\begin{align*}MD&= \displaystyle\frac{1}{n}\big(|x_1-\bar{x}|+|x_2-\bar{x}|+\ldots+|x_n-\bar{x}|\big)\\ &= h(w_1)\cdot|w_1-\bar{x}| + h(w_2)\cdot|w_2-\bar{x}| + \ldots + h(w_m)\cdot|w_m-\bar{x}|\;.\rule{0cm}{1cm}\end{align*}$$ Weil die Absolutbeträge technische Schwierigkeiten mit sich bringen, wird häufiger die **mittlere quadratische Abweichung** $\,\pmb{MQD}\,$ der Daten vom arithmetischen Mittel $\,\pmb{\bar{x}}\,$ $$\begin{align*}MQD&= \displaystyle\frac{1}{n}\big((x_1-\bar{x})^2+(x_2-\bar{x})^2+\ldots+(x_n-\bar{x})^2\big)\\ &= h(w_1)\cdot(w_1-\bar{x})^2+h(w_2)\cdot(w_2-\bar{x})^2+\ldots+h(w_m)\cdot(w_n-\bar{x})^2 \rule{0cm}{1cm}\end{align*}$$ gebraucht. Dieser Wert hat den Nachteil, dass die **Dimension** nicht mehr stimmt. Die Zahl $\,\pmb{MQD}\,$ **hat quadratische Einheiten**. Wenn wir aber aus $MQD$ die Wurzel zeihen, erhalten wir die sogenannte **Standardabweichung** $$s=\sqrt{MQD}\;,$$ **bei der die Dimension wieder stimmt**.     

Für **metrische Daten** stehen diese Lage- und Streumasse in Python **alle als Methoden** zur Verfügung. 

Für das metrische Merkmal **Alter** finden wir den **Median** $\,\pmb{x_{50}}\,$ mit der Methode **median()**:

In [15]:
q50 = df["Alter"].median()
print("Median =",q50) 

Median = 27.0


Das **arithmetische Mittel** $\,\pmb{\bar{x}}\,$ mit der Methode **mean()**:

In [16]:
x_bar = df["Alter"].mean()
print("arithmetisches Mittel =",x_bar) 

arithmetisches Mittel = 31.9


Mit der Methode **quantile($\,p/100\,$)** können wir beliebige $p$-**Quantile** bestimmen und damit also auch den Quartielsabstand $\,\pmb{Q}\,$:

In [17]:
Q = df["Alter"].quantile(0.75) - df["Alter"].quantile(0.25)
print("Q =",Q)

Q = 4.25


Die **mittlere Abweichung** $\,\pmb{MD}\,$ können wir mit den Methode **np.abs()** für den **Absolutbetrag** einer Zahl ausrechnen: 

In [18]:
MD = sum(np.abs(df["Alter"]-x_bar))/df.shape[0]
print("MD =",MD)

MD = 7.839999999999999


Für die **Standardabweichung** können wir die Methode **std()** brauchen. Dabei ist es wichtig, den Parameter **ddof** richtig zu setzen:
- Mit $\;\pmb{\texttt{ddof=0}}\;$ wird in der Berechung für $\,\pmb{MQD}\,$ **durch** $\,\textcolor{red}{\pmb{n}}\,$ **dividiert**
- Mit $\;\pmb{\texttt{ddof=1}}\;$ wird in der Berechung für $\,\pmb{MQD}\,$ **durch** $\,\textcolor{red}{\pmb{n-1}}\,$ **dividiert**

(Warum beide Methoden gebraucht werden, sehen wir später!) 

Für **die Standardabweichung** $\,\pmb{s}\,$ **gegebener Daten** setzen wir den Parameter immer auf $\;\textcolor{red}{\pmb{\texttt{ddof=0}}}\;$:

In [19]:
s = df["Alter"].std(ddof=0)
print("s =",s)

s = 12.676355943251199


***
Durch Einsetzen der Zahlen erhalten wir das selbe **von Hand**, ohne Python:

$\small\begin{align*}\bar{x}&= \frac{1}{10}(27+34+29+24+25+27+27+69+26+31) = 31.9\\
&= 0.1\cdot 24 + 0.1\cdot 25 + 0.1\cdot 26 + 0.3\cdot 27 + 0.1\cdot 29 + 0.1\cdot 31 + 0.1\cdot 34 + 0.1\cdot 69 = 31.9\rule{0cm}{0.7cm}     
\end{align*}$

$\small\begin{align*}MD&= \frac{1}{10}\big(|27-31.9|+|34-31.9|+|29-31.9|+|24-31.9|+|25-31.9|+|27-31.9|+|27-31.9|+|69-31.9|+|26-31.9|+|31-31.9|\big) = 7.84\\
&= 0.1\cdot|24-31.9| + 0.1\cdot|25-31.9| + 0.1\cdot|26-31.9| + 0.3\cdot |27-31.9| + 0.1\cdot |29-31.9| + 0.1\cdot|31-31.9| + 0.1\cdot |34-31.9| + 0.1\cdot|69-31.9| = 7.84\rule{0cm}{0.7cm}    
\end{align*}$

$\small\begin{align*}MQD&= \frac{1}{10}\big((27-31.9)^2+(34-31.9)^2+(29-31.9)^2+(24-31.9)^2+(25-31.9)^2+(27-31.9)^2+(27-31.9)^2+(69-31.9)^2+(26-31.9)^2+(31-31.9)^2\big) = 160.69\\
&= 0.1\cdot(24-31.9)^2 + 0.1\cdot(25-31.9)^2 + 0.1\cdot(26-31.9)^2 + 0.3\cdot (27-31.9)^2 + 0.1\cdot (29-31.9)^2 + 0.1\cdot(31-31.9)^2 + 0.1\cdot (34-31.9)^2 + 0.1\cdot(69-31.9)^2 = 160.69\rule{0cm}{0.7cm}\\
s &= \sqrt{160.69} \approx 12.676 \rule{0cm}{0.7cm}
\end{align*}$
***

### <span style="color:rgb(160,0,86)">Aufgabe 1</span>

Laden Sie die Daten *2021_Personalerhebung.csv* und bestimmen Sie für alle Merkmale passende Lagemasse und Streumasse. Bestimmen Sie auch für die Merkmale **Abteilung** und **Ausbildung** die relativen Häufigkeiten und relativen Summenhäufigkeiten.

***

### <span style="color:rgb(160,0,86)">Lösungsvorschlag</span>

**Teilaufgabe:**
- Wir laden die Daten in *2021_Personalerhebung.csv* und zeigen die ersten Zeilen an.  

In [20]:
import numpy as np
import pandas as pd

personal = pd.read_csv("Daten/2021_Personalerhebung.csv")
personal.head()

FileNotFoundError: [Errno 2] No such file or directory: 'Daten/2021_Personalerhebung.csv'

**Das sieht nicht gut aus!** 

Der **Seperator** in der Datei *2021_Personalerhabung.csv* ist **nicht das Komma** $(\,\pmb{,}\,)$ sonder das **Semikolon** $(\,\pmb{;}\,)$. In der Funktion **pd.read_csv()** können wir mit dem Parameter ***sep*** den Seperator angeben, der in der Datei die Spalten trennt:

In [ ]:
personal = pd.read_csv("Daten/2021_Personalerhebung.csv",sep=";")
personal.head()

,PersNummer,Unnamed: 1,Abteilung,Ausbildung,Eintrittsjahr,Bruttogehalt
0,560426,NaN,Finanzen,Abitur,2006,13200
1,590303,NaN,Vertrieb,Mittlere Reife,2008,13500
2,611117,NaN,Entwicklung,Promotion,2008,17600
3,620212,NaN,Geschaftsführung,Master,2006,18000
4,620624,NaN,Entwicklung,Master,2007,17400


Die Spalten **PersNummer** und **Unnamed: 1** brauchen wir nicht. Mit der Methode **drop()** können wir bei einem DataFrame Objekt **Spalten** (also mit dem Parameter ***axis=1***) **löschen**:  

In [ ]:
personal.drop(["PersNummer","Unnamed: 1"],axis=1,inplace=True)
personal.head()

,Abteilung,Ausbildung,Eintrittsjahr,Bruttogehalt
0,Finanzen,Abitur,2006,13200
1,Vertrieb,Mittlere Reife,2008,13500
2,Entwicklung,Promotion,2008,17600
3,Geschaftsführung,Master,2006,18000
4,Entwicklung,Master,2007,17400


***

**Teilaufgabe:**
- Wir bestimmen für das nominale Merkmal **Abteilung** die relativen Häufigkeiten.  

In [ ]:
werteAbteilung, absHAbteilung = np.unique(personal["Abteilung"],return_counts=True)

print("Verteilung der relativen Häufigkeiten:")
for i in range(len(werteAbteilung)):
    print("  h(" + werteAbteilung[i] +") =",absHAbteilung[i]/personal.shape[0])

Verteilung der relativen Häufigkeiten:
  h(Entwicklung) = 0.28
  h(Finanzen) = 0.12
  h(Geschaftsführung) = 0.16
  h(Schulung) = 0.12
  h(Test/Anwendungen) = 0.16
  h(Vertrieb) = 0.16


***

**Teilaufgabe:**
- Wir bestimmen für das ordinale Merkmal **Ausbildung** die relativen Summenhäufigkeiten.  

In [ ]:
werteAusbildung, absHAusbildung = np.unique(personal["Ausbildung"],return_counts=True)
werteAusbildung

array(['Abitur', 'Bache lor', 'Bachelor', 'Master', 'Mittlere Reife',
       'Promotion'], dtype=object)

Es gibt einen **Tipfehler** im Datensatz: Der Wert <tt>'Bache lor'</tt> muss korrigiert werden! 

Mit der Methode **replace()** können wir im ganzen DataFrame Objekt **personal** den falschen Wert ersetzen:

In [ ]:
personal = personal.replace("Bache lor","Bachelor")

werteAusbildung, absHAusbildung = np.unique(personal["Ausbildung"],return_counts=True)
print(werteAusbildung)
print(absHAusbildung)

['Abitur' 'Bachelor' 'Master' 'Mittlere Reife' 'Promotion']
[6 9 5 3 2]


Die **natürliche Reihenfolge** der Werte im Merkaml **Ausbildung** ist: 

<p style="text-align: center;"><tt>'Mittlere Reife'</tt>, <tt>'Abitur'</tt>, <tt>'Bachelor'</tt>, <tt>'Master'</tt>, <tt>'Promotion'</tt></p>

Für die weitere Arbeit speichern die Werte in den Variablen **werteAusbildung**  und **absHAusbildung** entsprechend:

In [ ]:
werteAusbildung = np.array([werteAusbildung[3],
                            werteAusbildung[0], 
                            werteAusbildung[1],
                            werteAusbildung[2],
                            werteAusbildung[4]])
absHAusbildung =  np.array([absHAusbildung[3],
                            absHAusbildung[0],
                            absHAusbildung[1],
                            absHAusbildung[2],
                            absHAusbildung[4]])
print(werteAusbildung)
print(absHAusbildung)

['Mittlere Reife' 'Abitur' 'Bachelor' 'Master' 'Promotion']
[3 6 9 5 2]


Mit diesen Angaben können wir die **Summenhäufigkeiten** bestimmen:

In [ ]:
print("Verteilung der relativen Summenhäufigkeiten:")
absSH = 0
for i in range(len(werteAusbildung)):
    absSH += absHAusbildung[i]
    print("  H(" + werteAusbildung[i] + ") =",absSH/personal.shape[0])     

Verteilung der relativen Summenhäufigkeiten:
  H(Mittlere Reife) = 0.12
  H(Abitur) = 0.36
  H(Bachelor) = 0.72
  H(Master) = 0.92
  H(Promotion) = 1.0


***

**Teilaufgabe:**
- Wir bestimmen bestimmen für alle Merkmale passende **Lagemasse** und **Streumasse**.  

Das Merkmal **Abteilung** ist **nominal**:
- die **Lage** der Daten können wir mit dem **Modus** $\,\pmb{\bar{x}_D}\,$ beschreiben
- die **Streuung** der Daten können wir mit dem **Dispersionsindex** $\,\pmb{P}\,$ beschreiben

In [ ]:
m = len(werteAbteilung) # Anzahl verschiedene Werte im Merkmal Abteilung

modus = personal["Abteilung"].mode()[0]
print("Der Modus ist:", modus)

relHAbteilung = absHAbteilung / personal.shape[0]
P = m/(m-1) * sum(relHAbteilung * (1-relHAbteilung))
print("Der Disperionsindex ist:",P)

Der Modus ist: Entwicklung
Der Disperionsindex ist: 0.9791999999999998


- Der Wert <tt>'Entwicklung'</tt> ist also representativ für das Merkmal **Abteilung**.
- Das Merkmal **Abteilung** hat eine hohe Dispersion. Die Verteilung der Werte ist demnach näher bei einer Gleichverteilung als bei einer Einpunkte-Verteilung. Das Merkmal **Abteilung** weist also eine starke Unterschiedlichkeit auf. 

Das Merkmal **Ausbildung** ist **ordinal**:
- die **mittlere Lage** der Daten können wir mit dem **Median** $\,\pmb{x_{50}}\,$ beschreiben
- die **Streuung** der Daten können wir mit der **Diversität** $\,\pmb{D}\,$ beschreiben

In [ ]:
m = len(werteAusbildung) # Anzahl verschiedene Werte im Merkmal Ausbildung

relHAusbildung = absHAusbildung / personal.shape[0]
relSHAusbildung = np.array([sum(relHAusbildung[0:i+1]) for i in range(m)])

i = 0
while i < m:
    if relSHAusbildung[i] > 0.5:
        break
    i += 1
median = werteAusbildung[i]
print("Der Median ist:",median)

D = 4/(m-1) * sum(relSHAusbildung * (1-relSHAusbildung))
print("Die Diversität ist:",D)

Der Median ist: Bachelor
Die Diversität ist: 0.6112000000000001


- Der Wert <tt>'Bachelor'</tt> beschreibt also die Mitte des Merkmals **Ausbildung**. Ungefähr 50% haben höchstens einen Bachelor und ungefähr 50% haben mindestens einen Bachelor. 
- Das Merkmal **Ausbildung** hat eine geringe bis mittlere Diversität. Die Verteilung der Werte gleicht demnach nicht einer extremen Zweipunkte-Verteilung (gleichmässig links und rechts verteilt) und auch nicht einer Einpunkte-Verteilung.  

Die Merkmale **Eintrittsjahr** und **Bruttogehalt** sind **metrisch**:

- die **mittlere Lage** der Daten können wir mit dem **Median** $\,\pmb{x_{50}}\,$ oder dem **arithmetischen Mittel** $\,\pmb{\bar{x}}\,$ beschreiben
- die **Streuung** der Daten können wir mit dem **Quartilsabstand** $\,\pmb{Q}\,$, der **mittleren Abweichung** $\,\pmb{MD}\,$ oder der **Standardabweichung** $\,\pmb{s}\,$ beschreiben

In [ ]:
medianEintritt, medianGehalt = personal[["Eintrittsjahr","Bruttogehalt"]].median()
print("Der Median beim Eintrittsjahr ist:",medianEintritt)
print("Der Median beim Gehalt ist:",medianGehalt,end="\n\n")

mittelEintritt, mittelGehalt = personal[["Eintrittsjahr","Bruttogehalt"]].mean()
print("Der arithmetische Mittelwert beim Eintrittsjahr ist:",mittelEintritt)
print("Der arithmetische Mittelwert beim Gehalt ist:",mittelGehalt,end="\n\n")

QEintritt, QGehalt = (personal[["Eintrittsjahr","Bruttogehalt"]].quantile(0.75) -
                      personal[["Eintrittsjahr","Bruttogehalt"]].quantile(0.25))
print("Der Quartilsabstand beim Eintrittsjahr ist:",QEintritt)
print("Der Quartilsabstand beim Gehalt ist:",QGehalt,end="\n\n")

MDEintritt = np.abs(personal["Eintrittsjahr"]-mittelEintritt).mean()
MDGehalt = np.abs(personal["Bruttogehalt"]-mittelGehalt).mean()
print("Die mittlere Abweichung beim Eintrittsjahr ist:",MDEintritt)
print("Der mittlere Abweichung beim Gehalt ist:",MDGehalt,end="\n\n")

sEintritt, sGehalt = personal[["Eintrittsjahr","Bruttogehalt"]].std(ddof=0)
print("Die Standardabweichung beim Eintrittsjahr ist:",sEintritt)
print("Die Standardabweichung beim Gehalt ist:",sGehalt)

Der Median beim Eintrittsjahr ist: 2010.0
Der Median beim Gehalt ist: 7200.0

Der arithmetische Mittelwert beim Eintrittsjahr ist: 2009.0
Der arithmetische Mittelwert beim Gehalt ist: 8840.0

Der Quartilsabstand beim Eintrittsjahr ist: 2.0
Der Quartilsabstand beim Gehalt ist: 8200.0

Die mittlere Abweichung beim Eintrittsjahr ist: 1.04
Der mittlere Abweichung beim Gehalt ist: 4276.8

Die Standardabweichung beim Eintrittsjahr ist: 1.296148139681572
Die Standardabweichung beim Gehalt ist: 5057.351085301474


- beim Merkmal **Eintrittsjahr** haben wir keine extremen Werte, der Median $x_{50}=2010$ und das arithmetische Mittel $\bar{x}=2009$ für die mittlere Lage unterscheiden sich kaum. Daher ist auch der Quartilsabstand $Q=2$ ungefähr doppelt so gross wie die mittlere Abweichung $MD=1.04$, die angibt, wie gross die Abweichung vom arithemtischen Mittel nach unten und nach oben mit Mittel beträgt. Die Standardabweichung $s\approx 1.30$ ist grösser als die mittlere Abweichung, weil grosse quadratische Abweichungen mehr ins Gewicht fallen. <br> Der Median sagt uns, dass ungefähr die Hälfte des Personal im Jahr $2010$ oder vorher ins Unternehmen eingetreten ist und die andere Hälfte im Jahr $2010$ oder später. Die Eintrittsjahre liegen eng beieinander. Gemäss Quartilsabstand liegen 50% der mittleren Eintrittsjahre nicht weiter als $2$ Jahre auseinander. Das zeigt auch die mittlere Abweichung vom arithmetischen Mittel, durchschnittlich ist das Eintrittjahr $2009$ mit einer mittleren Abweichung von plus oder minus $1.04$ Jahre.   
- beim Merkmal **Bruttogehalt** gibt es extreme Werte mit grossem Gehalt. Daher ist das arithmetische Mittel $\bar{x}=8840$ deutlich grösser als der Median $x_{50}=7200$ und der Quartilsabstand $8200$ ist grösser als die doppelte mittlere Abweichung $MD\approx 4276.8$. Auch hier sind die grossen Gehälter der Grund, dass die Standardabweichung $s\approx 5057.35$ grösser ist als die mittlere Abweichung.<br> Der Median sagt uns, dass ungefähr die Hälfte des Personal ein Gehalt von $7200$ oder weniger hat und die andere Hälfte ein Gehalt von $7200$ oder mehr. Die Gehälter liegen weit auseinander. Gemäss Quartilsabstand unterscheiden sich 50% der mittleren Gehälter bis zu $8200$. Das zeigt auch die mittlere Abweichung vom arithmetischen Mittel, durchschnittlich ist das Gehalt $8840$ mit einer mittleren Abweichung von plus oder minus $4276.8$.  

### <span style="color:rgb(160,0,86)">Aufgabe 2</span>

Laden Sie die Daten *Gesundheitskosten.csv* und bestimmen Sie für alle Merkmale passende Lagemasse und Streumasse. 

***

### <span style="color:rgb(160,0,86)">Lösungsvorschlag</span>

**Teilaufgabe:**
- Wir laden die Daten in *Gesundheitskosten.csv* und zeigen die ersten Zeilen an.  

In [ ]:
import numpy as np
import pandas as pd

kosten = pd.read_csv("Daten/Gesundheitskosten.csv")
kosten.head()

,Unnamed: 0,Krankenhäuser,Sozialmedizin,Arztpraxen,Zahnmedizin,Andere,Unterstützer,Detailhandel,Prävention,Staat,Versicherer,Importe
0,1995,12618.01,5626.22,5426.39,2678.79,1850.24,539.26,4291.59,523.88,873.51,1412.37,216.10
1,1996,13190.23,5983.41,5679.40,2731.91,1952.40,530.95,4459.82,543.27,864.67,1613.91,222.72
2,1997,13306.79,6218.51,5889.70,2752.02,1972.85,510.27,4646.23,554.69,810.88,1654.12,228.30
3,1998,13733.42,6517.62,6245.69,2787.40,2077.57,573.09,4777.10,598.37,815.25,1718.39,233.29
4,1999,14276.61,6635.99,6510.67,2787.08,2165.11,586.25,4933.82,622.52,861.53,1712.13,238.48


Die Spalte **Unnamed: 0** brauchen wir nicht, wir löschen sie mit der Methode **drop()**:

In [ ]:
kosten.drop(["Unnamed: 0"],axis=1,inplace=True)
kosten.head()

,Krankenhäuser,Sozialmedizin,Arztpraxen,Zahnmedizin,Andere,Unterstützer,Detailhandel,Prävention,Staat,Versicherer,Importe
0,12618.01,5626.22,5426.39,2678.79,1850.24,539.26,4291.59,523.88,873.51,1412.37,216.10
1,13190.23,5983.41,5679.40,2731.91,1952.40,530.95,4459.82,543.27,864.67,1613.91,222.72
2,13306.79,6218.51,5889.70,2752.02,1972.85,510.27,4646.23,554.69,810.88,1654.12,228.30
3,13733.42,6517.62,6245.69,2787.40,2077.57,573.09,4777.10,598.37,815.25,1718.39,233.29
4,14276.61,6635.99,6510.67,2787.08,2165.11,586.25,4933.82,622.52,861.53,1712.13,238.48


***

**Teilaufgabe:**
- Wir bestimmen bestimmen für alle Merkmale passende **Lagemasse** und **Streumasse**.    

Alle Merkmale sind **metrisch**:

- die Lage der Daten können wir mit dem **Median** $\,\pmb{x_{50}}\,$ oder dem **arithmetischen Mittel** $\,\pmb{\bar{x}}\,$ beschreiben
- die Streuung der Daten können wir mit dem **Quartilsabstand** $\,\pmb{Q}\,$, der **mittleren Abweichung** $\,\pmb{MD}\,$ oder der **Standardabweichung** $\,\pmb{s}\,$ beschreiben

Wir können alle Masse in einer Tabelle **df** (ein DataFrame Objekt) zusammenstellen und mit der Funktion **np.round()** die Ergebnisse runden:

In [ ]:
df = pd.DataFrame({"Median":kosten.median(),
                   "Mittelwerte":kosten.mean(),
                   "Quartilsabstand":kosten.quantile(0.75)-kosten.quantile(0.25),
                   "Mittlere Abweichung":np.abs(kosten-kosten.mean()).mean(),
                   "Standardabweichung":kosten.std(ddof=0)})
np.round(df,1)

,Median,Mittelwerte,Quartilsabstand,Mittlere Abweichung,Standardabweichung
Krankenhäuser,20042.5,21013.2,9616.2,5088.8,5807.8
Sozialmedizin,9744.1,9855.7,4575.8,2312.4,2616.1
Arztpraxen,8719.3,9031.1,3969.7,2051.9,2383.2
Zahnmedizin,3670.0,3550.5,1049.5,523.6,579.6
Andere,3179.8,3529.3,1939.5,1146.3,1355.3
Unterstützer,850.2,982.1,660.4,350.8,414.0
Detailhandel,6198.3,6087.6,1306.8,802.5,950.7
Prävention,818.0,821.0,333.4,159.5,179.5
Staat,1244.1,1187.5,334.8,227.1,324.1
Versicherer,2200.4,2156.2,690.7,366.4,411.5


### <span style="color:rgb(160,0,86)">Aufgabe 3</span>

Laden Sie die Daten *2022_Mathematik 1 WiSo Urliste*. In dieser Excel Datei sind die Mathematik Prüfungsergebnisse von Studierenden an der Fakultät Wirschaft und Sozialwissenschaften der Universität Bern gespeichert. Alle Studierende haben die gleiche Prüfung geschrieben, nur die Reihenfolge der Aufgaben war in der Version A und Version B anders (siehe Angaben in den Zellen N3 und N4). Die Punkte für die 9 Aufgaben sind abhängig von der Version A und B also in den Spalten vertauscht!  Bestimmen Sie für jede Aufgabe passende Lagemasse und Streumasse für die Punkte.  

***

### <span style="color:rgb(160,0,86)">Lösungsvorschlag</span>

Zuerst laden wir die Excel-Datei und speichern die Daten im DataFrame **punkte**. Die benötigten Werte müssen wir nur aus der Excel Datei herauslesen. Wir brauchen nur die Punkte zu den 9 Aufgaben und die Angabe zur Version der Prüfung. Zur besseren Übersicht können wir auch die Spalten einfacher benennen. Dann zerlegen wir das DataFrame **punkte** in zwei Teile:
-  **VersionA** enthält alle Studierenden, bei denen die Punkte zu den Aufgaben bereits in korrekter Reihenfolge gespeichert sind
-  **VersionB** enthält alle Studierenden, bei denen wir die Aufgaben A1 und A2, A3 und A4, A5 und A6, A7 und A8 vertauschen müssen. Nach dem Vertauschen bennen wir die Spalten mit den Aufgabennummern der Version A, weil diese die gleichen Aufaben sind.

Dann fügen wir die zwei Teile **VersionA** und **VersionB** wieder zu einem DataFrame **punkteNeu** zusammen.

In [ ]:
# pip install xlrd
punkte = pd.read_excel("Daten/2022_Mathematik 1 WiSo Urliste.xls")

# wir speichern die bestehenden Namen der Spalten
colNamen = punkte.columns

# und schneiden die relevanten Spalten (axis=1) und Zeilen (axis=0) heraus
punkte.drop(list(colNamen[0:1])+list(colNamen[10:11])+list(colNamen[12:]),axis=1,inplace=True)
punkte.drop([0,1,2,3],axis=0,inplace=True)

# nun geben wir den Spalten einfache Namen
colNamenNeu = ["A1","A2","A3","A4","A5","A6","A7","A8","A9","Version"]
punkte.columns = colNamenNeu

# dann speichern wir die Studierenden mit Veresion A
VersionA = punkte[colNamenNeu][punkte["Version"]=="A"]

# bei den Studierenden mit Version B müssen wir die Reihenfolge der Aufgaben ändern
VersionB = punkte[["A2","A1","A4","A3","A6","A5","A8","A7","A9","Version"]][punkte["Version"]=="B"]
VersionB.columns = colNamenNeu

# zuletzt fügen wir die zwei DataFrame VersionA und VersionB wieder zusammen
punkteNeu = pd.concat([VersionA,VersionB])
punkteNeu

,A1,A2,A3,A4,A5,A6,A7,A8,A9,Version
4,3.5,4,3.5,4,4,4,4,4,4,A
6,1.5,4,3,2.5,3,4,1.5,3,2,A
7,1.5,3.5,1.5,4,3,0,1,3.5,0,A
8,3.5,3,3.5,3,3.5,2,3.5,2.5,2,A
10,4,3.5,3.5,4,3,4,3.5,3,3.5,A
...,...,...,...,...,...,...,...,...,...,...
405,1,1.5,1.5,1,2,0,0.5,1.5,0,B
406,0,1,0,0,0.5,0,0.5,0,0,B
408,3.5,3.5,4,3,2.5,4,4,4,4,B
411,2.5,3,1.5,2,3,0.5,1,1,0,B


Die Werte der 9 Aufgaben sind alle **metrisch**:

- die Lage der Daten können wir mit dem **Median** $x_{50}$ oder dem **arithmetischen Mittel** $\bar{x}$ beschreiben
- die Streuung der Daten können wir mit dem **Quartilsabstand** $Q$, der **mittleren Abweichung** $MD$ oder der **Standardabweichung** $s$ beschreiben

Die Spalte **Version** brachen wir für die Berechnungen nicht. Wir erstellen ein neues DataFrame **pkt**, das nur die Punkte zu den 9 Aufgaben enthält.

Wir stellen wieder alle Masse in einer Tabelle **df** (ein DataFrame Objekt) zusammen und runden die Werte mit der Funktion **np.round()**:

In [ ]:
pkt = punkteNeu[colNamenNeu[0:9]]

df = pd.DataFrame({"Median":pkt.median(),
                   "Mittelwerte":pkt.mean(),
                   "Quartilsabstand":pkt.quantile(0.75)-pkt.quantile(0.25),
                   "Mittlere Abweichung":np.abs(pkt-pkt.mean()).mean(),
                   "Standardabweichung":pkt.std(ddof=0)})
np.round(df,1)

,Median,Mittelwerte,Quartilsabstand,Mittlere Abweichung,Standardabweichung
A1,2.0,2.155718,2.5,1.187366,1.282992
A2,3.0,2.868613,1.0,0.755821,0.92991
A3,2.5,2.411192,2.0,1.05203,1.242014
A4,3.0,2.604623,2.5,1.182174,1.354457
A5,3.0,2.821168,2.0,0.913279,1.154812
A6,1.0,1.504866,1.5,0.998479,1.209241
A7,2.5,2.30292,2.5,1.303959,1.416394
A8,2.5,2.164234,2.5,1.151846,1.318652
A9,0.5,1.069343,2.0,1.093754,1.251543


![HSLU](Bilder/LogoHSLU.png)